src/intrinsics/ecma262/Object.js

```diff
@@ -73,12 +74,46 @@ export default function(realm: Realm): NativeFunctionValue {
       sources;
       let delayedSources = [];
 
-      // 4. For each element nextSource of sources, in ascending index order,
-      for (let nextSource of sources) {
-        let keys, frm, frmSnapshot;
+      const handleSnapshot = (frm, frm_was_partial) => {
+        if (to_must_be_partial) {
+          if (to instanceof AbstractObjectValue && to.values.isTop()) {
+            // We don't know which objects to make partial and making all objects partial is failure in itself
+            AbstractValue.reportIntrospectionError(to);
+            throw new FatalError();
+          } else {
+            // if to has properties, we better remove them because after the temporal call to Object.assign we don't know their values anymore
+            if (to.hasStringOrSymbolProperties()) {
+              // preserve them in a snapshot and add the snapshot to the sources
+              delayedSources.push(to.getSnapshot({ removeProperties: true }));
+            }
+
+            if (frm instanceof ObjectValue && frm.mightBeHavocedObject()) {
+              // it's not safe to trust any of its values
+              let temporal = AbstractValue.createTemporalFromBuildFunction(realm, ObjectValue, [frm], ([x]) => x, {
```

Comment:
I don't understand what's going on here.

Why isn't this just `delayedSources.push(frm)`?